service/history/queueFactoryBase.go

```diff
@@ -87,33 +88,60 @@ type (
 		fx.In
 
 		Lifecycle fx.Lifecycle
-		Factories []QueueFactory `group:"queueFactory"`
+		Factories []QueueFactory
 	}
 )
 
 var QueueModule = fx.Options(
 	fx.Provide(QueueSchedulerRateLimiterProvider),
 	fx.Provide(
 		fx.Annotated{
-			Group:  QueueFactoryFxGroup,
+			Name:   "transferQueueFactory",
 			Target: NewTransferQueueFactory,
 		},
 		fx.Annotated{
-			Group:  QueueFactoryFxGroup,
+			Name:   "timerQueueFactory",
 			Target: NewTimerQueueFactory,
 		},
 		fx.Annotated{
-			Group:  QueueFactoryFxGroup,
+			Name:   "visibilityQueueFactory",
 			Target: NewVisibilityQueueFactory,
 		},
 		fx.Annotated{
-			Group:  QueueFactoryFxGroup,
+			Name:   "archivalQueueFactory",
 			Target: NewArchivalQueueFactory,
 		},
+		getQueueFactories,
 	),
 	fx.Invoke(QueueFactoryLifetimeHooks),
 )
 
+type queueFactorySet struct {
+	fx.In
+
+	TransferQueueFactory   QueueFactory `name:"transferQueueFactory"`
+	TimerQueueFactory      QueueFactory `name:"timerQueueFactory"`
+	VisibilityQueueFactory QueueFactory `name:"visibilityQueueFactory"`
+	ArchivalQueueFactory   QueueFactory `name:"archivalQueueFactory"`
+}
+
+// getQueueFactories returns factories for all the enabled queue types.
+// The archival queue factory is only returned when archival is enabled in the static config.
+func getQueueFactories(
+	queueFactorySet queueFactorySet,
+	archivalMetadata archiver.ArchivalMetadata,
+) []QueueFactory {
+	factories := []QueueFactory{
+		queueFactorySet.TransferQueueFactory,
+		queueFactorySet.TimerQueueFactory,
+		queueFactorySet.VisibilityQueueFactory,
+	}
+	if archivalMetadata.GetHistoryConfig().StaticClusterState() == archiver.ArchivalEnabled || archivalMetadata.GetVisibilityConfig().StaticClusterState() == archiver.ArchivalEnabled {
```

Comment:
can it call ClusterConfiguredForArchival?

Reply:
No because that can change based on the dynamic config, and so we need to turn it on during startup using the static config because otherwise the dynamic config for durable archival may turn on, and then we'll start producing tasks to a queue which is not running: 
<img width="593" alt="image" src="https://user-images.githubusercontent.com/5942963/214166370-683cb6c8-fe1b-499c-9622-272688a39a67.png">

Reply:
<img width="574" alt="image" src="https://user-images.githubusercontent.com/5942963/214166454-d65fa330-b9d2-4e6c-ac7e-d20601148654.png">

Reply:
Make sense to me now. I guess this will be simplified after deprecate the archival workflow?

Reply:
Unfortunately, not really. The durable archival flag will be removed, but the two-tier archival switch will still exist, and we actually need to keep it that way.